### PR TITLE
bridge: add vlan trunk support

### PR DESCRIPTION
**Pull request by tjjh89017**
add vlan trunk support for veth
vlan trunk only support L2 only mode without any IPAM
refer ovs-cni design
https://github.com/k8snetworkplumbingwg/ovs-cni/blob/main/pkg/plugin/plugin.go

design:
origin "vlan" option will be PVID or untagged vlan for the network.
"vlanTrunk" will setup tagged vlan for veth.

entry type:
`{ "id": 100 }` will specify only tagged vlan 100
`{ "minID": 100, "maxID": 120 }` will specify tagged vlan from 100 to
120 (include 100 and 120)
vlanTrunk is a list of above entry type, so you can use this to add
tagged vlan
`[
  { "id": 100 },
  {
    "minID": 1000,
    "maxID": 2000
  }
]`

complete config will be like this
```json
{
  "cniVersion": "0.3.1",
  "name": "mynet",
  "type": "bridge",
  "bridge": "mynet0",
  "vlan": 100,
  "vlanTrunk": [
    { "id": 101 },
    { "minID": 1000, "maxID": 2000 },
    { "minID": 3000, "maxID": 4000 }
  ],
  "ipam": {}
}
```

Signed-off-by: Date Huang <date.huang@suse.com>

### CODE REVIEW

**Comment by tjjh89017** [comment]
@dcbw could you take a look on it.
thanks a lot!

File: plugins/main/bridge/bridge.go [hunk 1]
```diff
@@ -373,6 +440,16 @@ func setupVeth(netns ns.NetNS, br *netlink.Bridge, ifName string, mtu int, hairp
 		}
 	}
 
+	if vlans != nil {
+		// TODO setup vlan trunk
+		for _, v := range vlans {
+			err = netlink.BridgeVlanAdd(hostVeth, uint16(v), false, false, false, true)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to setup vlan tag on interface %q: %v", hostIface.Name, err)
```
Comment (jonyhy96) [inline plugins/main/bridge/bridge.go:448]: may better to use `%w` rather than %v for err

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:448]: let me check the difference between %w and %v and feedback to you
tks for comment

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:448]: should we consider backward compatible to earlier golang?
if no, i will change to `%w`

File: plugins/main/bridge/bridge.go [hunk 2]
```diff
@@ -373,6 +440,16 @@ func setupVeth(netns ns.NetNS, br *netlink.Bridge, ifName string, mtu int, hairp
 		}
 	}
 
+	if vlans != nil {
+		// TODO setup vlan trunk
```
Comment (squeed) [inline plugins/main/bridge/bridge.go:444]: Is this still a TODO?

**Comment by squeed** [comment]
This is interesting. We had a question in the maintainers' meeting: how do you receive the packets inside the container? Are you using another process to create vlan-tagged interfaces inside? Or are you doing it in userspace?

**Comment by squeed** [comment]
Another comment (mostly from @dcbw) -- is trunking a per-port setting? Is it possible / desired for different containers on the same bridge to have different trunking configurations?

File: plugins/main/bridge/bridge.go (hunk 2 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:444]: it seems i forgot to remove it.

**Comment by tjjh89017** [comment]
> This is interesting. We had a question in the maintainers' meeting: how do you receive the packets inside the container? Are you using another process to create vlan-tagged interfaces inside? Or are you doing it in userspace?

In our scenario, kubevirt will setup another bridge to catch all traffic and let VM handle the VLAN tag
userspace program also could handle vlan tag with `NET_RAW`.
we also create vlan interface to filter vlan tag also.

> Another comment (mostly from @dcbw) -- is trunking a per-port setting? Is it possible / desired for different containers on the same bridge to have different trunking configurations?
Per port, yes!

linux equivalent command:
```bash
ip link add br0 type bridge vlan_filtering 1
ip link set br0 up
ip link set eth0 master br0
ip link add veth-red type veth peer name veth-blue
ip link set veth-red master br0
ip link set veth-red up
ip link add veth-green veth peer name veth-purple
ip link set veth-green master br0
ip link set veth-green up

birdge vlan add dev veth-red vid 100 pvid untagged
bridge vlan add dev veth-red vid 101
bridge vlan add dev veth-red vid 102

bridge vlan add dev veth-green vid 200 pvid untagged
bridge vlan add dev veth-green vid 201
bridge vlan add dev veth-green vid 202
```

File: plugins/main/bridge/bridge.go (hunk 1 above)
Comment (dcbw) [reply plugins/main/bridge/bridge.go:447]: @tjjh89017 yes, please use %w. CNI follows current golang (1.17) and previous golang (1.16) which both have it. We are just a bit lazy in regularly bumping requirements.

**Comment by dcbw** [comment]
@tjjh89017 could you rebase since we merged a fix for ipamDel()? Thanks!

File: plugins/main/bridge/bridge.go [hunk 3]
```diff
@@ -377,6 +444,15 @@ func setupVeth(netns ns.NetNS, br *netlink.Bridge, ifName string, mtu int, hairp
 		}
 	}
 
+	if vlans != nil {
+		for _, v := range vlans {
+			err = netlink.BridgeVlanAdd(hostVeth, uint16(v), false, false, false, true)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to setup vlan tag on interface %q: %v", hostIface.Name, err)
```
Comment (tjjh89017) [inline plugins/main/bridge/bridge.go:451]:
```suggestion
				return nil, nil, fmt.Errorf("failed to setup vlan tag on interface %q: %w", hostIface.Name, err)
```

**Comment by tjjh89017** [comment]
@dcbw tks for help
already rebased it.
need your double chceck again and approve.

File: plugins/main/bridge/bridge.go (hunk 1 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go]: changed to `%w` 
thanks for suggestion

**Comment by tjjh89017** [comment]
@dcbw do we have any plan to support this feature?
thanks a lot

File: plugins/main/bridge/bridge.go [hunk 4]
```diff
@@ -101,6 +110,11 @@ func loadNetConf(bytes []byte, envArgs string) (*NetConf, string, error) {
 	if n.Vlan < 0 || n.Vlan > 4094 {
 		return nil, "", fmt.Errorf("invalid VLAN ID %d (must be between 0 and 4094)", n.Vlan)
 	}
+	var err error
+	if n.vlans, err = collectVlanTrunk(n.VlanTrunk); err != nil {
```
Comment (austinvazquez) [inline plugins/main/bridge/bridge.go:114]: Would it be more consistent with the rest of the code to use the scoped variable and not define err?

File: plugins/main/bridge/bridge.go [hunk 5]
```diff
@@ -124,6 +138,59 @@ func loadNetConf(bytes []byte, envArgs string) (*NetConf, string, error) {
 	return n, n.CNIVersion, nil
 }
 
+// This method is copied from https://github.com/k8snetworkplumbingwg/ovs-cni/blob/main/pkg/plugin/plugin.go
```
Comment (austinvazquez) [inline plugins/main/bridge/bridge.go:141]: Minor change. It may be useful to pin this link to a release version instead of following main.

**Comment by tjjh89017** [comment]
@dcbw @squeed Hi 
Our project will still need this enhancement.
Could you guys give us some comment for this feature?
thanks a lot

**Comment by bk201** [comment]
Please help reopen the issue. We still need this enhancement.

File: plugins/main/bridge/bridge.go (hunk 4 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:114]: We will still need to copy the value back to `n.vlans` later.
So I will consider to keep current code

File: plugins/main/bridge/bridge.go (hunk 5 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:141]: good idea
thanks a lot

**Comment by squeed** [comment]
Arg, I somehow can't reopen. I'm sorry. Can you resubmit?

**Comment by tjjh89017** [comment]
@squeed I will create a new PR for this commit.
I will link to here later.

**Pull request closed without merging**